Add Discord integration

AuroraEchoes opened the pull request:
Add Discord integration to Citadel.

Discord integration can be toggled with the property `discord_integration` in `config/features.yml`.

# UI Changes
- Create “Connections” panel in Settings, allowing users to link a Discord account through OAuth.
- Add a hyperlink to the linked Discord account next to the Steam account link on the User show page.

# API Changes
- GET `/users/:id`: add `discord_id` as a 64-bit integer field.
- GET `/users/steam_id/:steam_id`: add `discord_id` as a 64-bit integer field.
- GET `/users/discord_id/:discord_id`: create endpoint. Will return the same json as /users/:id and /users/steam_id/:steam_id. Will return 403: Forbidden if `discord_integration` is disabled.
For more information on API changes, see `docs/api/v1.md`.

# Configuration Changes
- Add `discord_client_id` and `discord_client_secret` credentials.
- Create `features.yml`.

# Dependency Changes
- Add `omniauth-discord`


Added code should be entirely covered by tests. The GitHub Actions build failed because of a single minor Rubocop linting suggestion.

Thank you for taking the time to review my PR. Please don’t hold back on criticism; I’m happy to make changes.

AuroraEchoes committed:
Fix steam_id being passed as an integer

AuroraEchoes committed:
Add Discord account linking integration

AuroraEchoes committed:
Update Discord logo to use SVG version

AuroraEchoes committed:
Add tests for Discord integration

AuroraEchoes committed:
Add discord_client_id and discord_client_secret to workflows

AuroraEchoes committed:
Make Discord integration toggleable

AuroraEchoes committed:
Remember to add features.yml

AuroraEchoes committed:
Add unit tests for (hopefully) full coverage of Discord integration

AuroraEchoes committed:
Use a validation rule for enforcing Discord ID uniqueness

BenjaminSchaaf commented on app/models/user.rb:
```diff
@@ -106,11 +108,11 @@ def authorized_teams_for(league)
 
   def steam_32
-    SteamId.to_32(steam_id)
+    SteamId.to_32(steam_id.to_s)
```
I don't see why this is needed? The tests cover this case already. If anything the improvement here would be to use the `from_64_to_*` functions.

AuroraEchoes replied:
In my testing environment, this would cause an error because `steam_id` is stored in the database as a integer, and `SteamId.to_*` is expecting a string. Locally, all the `lib/steam_id` tests fail too.

That said, changing it to use `from_64_to_*` fixed the error for me in the case of the actual app; the tests are still failing.

BenjaminSchaaf replied:
Odd. Tests are working fine on my end. Are you perhaps using a newer version of ruby? citadel targets 3.1.2.

AuroraEchoes replied:
Yep, I’m running whatever `pacman` gives me, so, 3.4.3. I’ll switch my dev environment to use 3.1.2 in future!

BenjaminSchaaf commented on Gemfile.lock:
```diff
@@ -493,7 +521,7 @@ GEM
 
 PLATFORMS
-  ruby
+  x86_64-linux
```
I don't see why we're limiting to linux here.

AuroraEchoes replied:
Sorry, I didn’t change that but I presume a command I ran did. Sloppy on my part. Fixed in the patch.

BenjaminSchaaf commented on app/controllers/users_controller.rb:
```diff
@@ -91,6 +92,34 @@ def request_name_change
     end
   end
 
+  def link_discord
```
This is an unprotected method, allowing anyone to link anyone's discord to their account. All account linking needs to happen in `OmniauthCallbacksController#discord`, as that is the only place where we've actually verified the login.

AuroraEchoes replied:
Extremely good point! I moved the `link_discord` method to `OmniauthCallbacksController`. I assume that `unlink_discord` is still fine; it seems to be from my testing.

BenjaminSchaaf commented on app/models/user.rb:
```diff
@@ -34,11 +34,13 @@ class User < ApplicationRecord
   has_many :logs, class_name: 'User::Log', dependent: :destroy
 
-  devise :rememberable, :omniauthable, omniauth_providers: [:steam]
+  devise :rememberable, :omniauthable, omniauth_providers: [:steam, :discord]
```
Devise should not have the discord provider if the feature is disabled.

BenjaminSchaaf commented on app/presenters/user_presenter.rb:
```diff
@@ -40,6 +40,22 @@ def steam_icon_link(options = { size: 16 })
   end
 
+  def discord_profile_url
+    "http://discord.com/users/#{user.discord_id}"
```
`https`

BenjaminSchaaf commented on db/migrate/20250423113545_add_discord_id_to_users.rb:
```diff
@@ -0,0 +1,5 @@
+class AddDiscordIdToUsers < ActiveRecord::Migration[7.1]
+  def change
+    add_column :users, :discord_id, :bigint
```
Since we're looking up users by their discord id this column needs an index.

BenjaminSchaaf reviewed (changes requested):
Thanks for the PR. There's one major issue, but this otherwise looks pretty good.

AuroraEchoes committed:
Minor fixes: protect link_discord, only add Discord OmniAuth provider if integration is enabled

BenjaminSchaaf commented on db/migrate/20250423113545_add_discord_id_to_users.rb:
```diff
@@ -0,0 +1,6 @@
+class AddDiscordIdToUsers < ActiveRecord::Migration[7.1]
+  def change
+    add_column :users, :discord_id, :bigint
+    add_index :users, :discord_id
```
Should be `unique: true`

BenjaminSchaaf commented on config/routes.rb:
```diff
@@ -108,6 +109,9 @@
   resources :users, except: [:destroy] do
     post 'name',  on: :member, to: 'users#request_name_change'
+    member do
+      patch 'unlink_discord'
```
Should be `patch 'unlink_discord', on: :member`, or move all member routes into `member do`.

BenjaminSchaaf commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -16,6 +18,37 @@ def steam
+    end
+
+    def link_discord(user_id, discord_id)
+      if discord_integration_enabled?
```
This should really be checked before adding the discord routes. It shouldn't be possible to initiate a discord auth flow to even get to this point.

BenjaminSchaaf commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -16,6 +18,37 @@ def steam
+      link_discord(user_id, discord_id)
+    end
+
+    def link_discord(user_id, discord_id)
```
This should be private.

BenjaminSchaaf commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -16,6 +18,37 @@ def steam
+        if @user.update(discord_id: discord_id)
+          flash[:notice] = 'Discord account linked!'
+        else
+          flash[:error] = 'This Discord account is already linked to another Ozfortress account.'
```
Remove ozfortress.

BenjaminSchaaf commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -16,6 +18,37 @@ def steam
+    def discord
+      auth = request.env['omniauth.auth']
+      params = request.env['omniauth.params']
+      user_id = params['user_id']
```
`user_id` is client-provided, and thus cannot be trusted. `current_user` already refers to the currently logged in user that we want to link to.

BenjaminSchaaf commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -16,6 +18,37 @@ def steam
+      end
+    end
+
+    def discord_revoke_token(token)
```
Should be private.

BenjaminSchaaf commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -16,6 +18,37 @@ def steam
+      params = request.env['omniauth.params']
+      user_id = params['user_id']
+      discord_id = auth.extra.raw_info['id']
+      discord_revoke_token(auth.credentials.token)
```
I'm curious what the reason behind revoking the token is?

BenjaminSchaaf commented on app/controllers/users_controller.rb:
```diff
@@ -91,6 +92,18 @@ def request_name_change
 
+  def unlink_discord
+    if discord_integration_enabled?
+      @user = User.find(params[:id])
```
The first before_action does this already. Add `unlink_discord` to it.

BenjaminSchaaf commented on app/controllers/users_controller.rb:
```diff
@@ -91,6 +92,18 @@ def request_name_change
     end
   end
 
+  def unlink_discord
```
We need to require user permission for this. Otherwise anyone can unlink anyone's discord.

BenjaminSchaaf commented on app/controllers/users_controller.rb:
```diff
@@ -91,6 +92,18 @@ def request_name_change
   end
 
+  def unlink_discord
+    if discord_integration_enabled?
```
This route shouldn't be accessible if discord is disabled so the check shouldn't happen here.

BenjaminSchaaf commented on app/views/users/edit.html.haml:
```diff
@@ -15,6 +15,23 @@
+        - if @user.discord_id
+          .flex-fill
+            Account Linked
+          = button_to 'Unlink Account', unlink_discord_user_path(@user), method: :patch, class: 'btn btn-danger'
```
This needs a confirmation dialog.

BenjaminSchaaf commented on app/views/users/show.html.haml:
```diff
@@ -24,7 +24,12 @@
       %span.mx-1
-      = user_p.steam_id_link
+        = user_p.steam_id_link
+      - if discord_integration_enabled? && @user.discord_id?
```
This should certainly go after all the steam links, maybe on its own row.

BenjaminSchaaf commented on app/presenters/user_presenter.rb:
```diff
@@ -40,6 +40,22 @@ def steam_icon_link(options = { size: 16 })
+  end
+
+  def discord_icon_link(options = { size: 16 })
+    link_to(discord_id_link, target: '_blank', rel: 'noopener') do
```
The first argument to link_to must be a url. You're passing in a HTML <a> element.

BenjaminSchaaf reviewed (changes requested):
Second round of review, some good progress.

AuroraEchoes committed:
Don’t register Discord routes unless the feature is enabled, review round 2 fixes

AuroraEchoes commented:
Round 2 revisions should be in! A couple of notes:
- Revoking the token isn’t strictly necessary, but it seems like good practice. We’re not going to use it again, and it has a lifetime of 14 days, so it seems better to dispose of it.
- Having `Features` as a concern is looking increasingly sketchy. Especially because we keep on needing to access it statically, so have to do the long route (`Rails.configuration.features[:discord_integration]`). I think that it makes sense to not just read from the Rails configuration, both because of how long the path is, and because future uses of `Features` could require some processing of the feature configuration. What jumps out at me is a singleton, but because (AFAIK) there are no other singletons in the codebase, asking for guidance seems prudent.
- In `routes.rb`, if Discord integration isn’t enabled, the `unlink_discord` route isn’t registered (using an `if` statement as opposed to a constraint). This made more sense to me, as it is consistent with the approach for the Discord OAuth routes — the routes are never registered if Discord integration is disabled. However, I understand this to be slightly non-idiomatic, so again, not confident on the preferred approach here.

BenjaminSchaaf commented on config/routes.rb:
```diff
@@ -10,6 +10,7 @@
       end
 
       get 'users/steam_id/:id', to: 'users#steam_id'
+      get 'users/discord_id/:id', to: 'users#discord_id'
```
Should also have a check for the feature.

BenjaminSchaaf commented on app/controllers/api/v1/users_controller.rb:
```diff
@@ -12,6 +14,16 @@ def steam_id
       end
+
+      def discord_id
+        if discord_integration_enabled?
```
This should be inaccessible if the feature is disabled, making the check redundant.

BenjaminSchaaf commented on app/controllers/users_controller.rb:
```diff
@@ -91,6 +92,12 @@ def request_name_change
   end
 
+  def unlink_discord
+    @user.update(discord_id: nil)
```
`update!`

BenjaminSchaaf commented on config/database.yml:
```diff
@@ -2,6 +2,10 @@ default: &default
   adapter: postgresql
   encoding: utf-8
   pool: 5
+  username: citadel_dev
```
This shouldn't be committed.

AuroraEchoes replied:
Oops!

BenjaminSchaaf commented on spec/controllers/users/omniauth_callback_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+    it 'allows users to link a Discord account' do
+      request.env['devise.mapping'] = Devise.mappings[:user]
+      request.env['omniauth.auth'] = OmniAuth.config.mock_auth[:discord]
+      request.env['omniauth.params'] = { 'user_id' => user.id }
```
Don't think this is used anymore

BenjaminSchaaf commented on spec/controllers/users/omniauth_callback_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+    it 'enforces Discord account uniqueness' do
+      request.env['devise.mapping'] = Devise.mappings[:user]
+      request.env['omniauth.auth'] = OmniAuth.config.mock_auth[:discord]
+      request.env['omniauth.params'] = { 'user_id' => user.id }
```
ditto

BenjaminSchaaf commented on spec/controllers/users/omniauth_callback_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+      request.env['omniauth.auth'] = OmniAuth.config.mock_auth[:discord]
+      request.env['omniauth.params'] = { 'user_id' => user.id }
+      user_prev_id = user.discord_id
+      second_user.update(discord_id: discord_id)
```
`update!`

BenjaminSchaaf commented on spec/controllers/users/omniauth_callback_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+      request.env['omniauth.params'] = { 'user_id' => user.id }
+      user_prev_id = user.discord_id
+      second_user.update(discord_id: discord_id)
+      second_user.reload
```
Reloading is not needed here. We don't use `second_user` again, nor is it mutated externally.

BenjaminSchaaf commented on spec/controllers/users/omniauth_callback_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+    let(:discord_id) { 123 }
+
+    before do
+      Rails.configuration.features[:discord_integration] = true
```
We should use a mock instead:

```
allow(Rails.configuration.features).to receive(:[]).with(:discord_integration).and_return(true)
```

This will automatically clean up afterwards.

BenjaminSchaaf commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -360,4 +360,31 @@
     end
   end
+
+  describe 'PATCH #unlink_discord' do
```
Missing tests for permissions.

BenjaminSchaaf commented on spec/requests/api/v1/users_spec.rb:
```diff
@@ -75,4 +75,46 @@
+
+    it 'succeeds for existing user' do
+      Rails.configuration.features[:discord_integration] = true
+      user.update(discord_id: 123)
```
`update!`

BenjaminSchaaf commented on spec/views/users/show.html.haml_spec.rb:
```diff
@@ -49,6 +50,9 @@
     expect(rendered).to include(user.name)
     expect(rendered).to include(user.badge_name)
+    unless user.discord_id.nil? || !discord_integration_enabled?
+      expect(rendered).to include(present(user).discord_id_link)
```
Tests shouldn't rely on configuration like this. There should be two tests, with discord integration enabled and disabled respectively.

Additionally we already know the user has a discord id since we used `:user_with_discord`. There's no need to check.

BenjaminSchaaf commented on spec/views/users/show.html.haml_spec.rb:
```diff
@@ -59,6 +63,9 @@
 
     expect(rendered).to include(user.name)
     expect(rendered).to include(user.badge_name)
+    unless user.discord_id.nil? || !discord_integration_enabled?
```
ditto

BenjaminSchaaf reviewed (changes requested):
The approach is good now. There's mostly minor stuff left. CI also has a failure.

AuroraEchoes committed:
Unit test improvements, disable API routes if integration disabled

AuroraEchoes committed:
Rebase feature/discord-integration onto master

AuroraEchoes committed:
Fix Gemfile versioning

AuroraEchoes committed:
Rubocop autocorrect

AuroraEchoes committed:
Revert Gemfile changes

BenjaminSchaaf commented on Gemfile.lock:
```diff
@@ -287,12 +299,23 @@ GEM
     nio4r (2.7.3)
-    nokogiri (1.16.5)
-      mini_portile2 (~> 2.8.2)
+    nokogiri (1.16.5-x86_64-linux)
```
Not sure how this happened, but that's definitely not right.

AuroraEchoes replied:
Bundler loves doing this on every run without telling me, it turns out.

BenjaminSchaaf commented on app/views/users/show.html.haml:
```diff
@@ -21,14 +21,19 @@
       %span= @user.steam_32
       %span.mx-2= '/'
       %span= @user.steam_id
+    .row.align-items-center.external-link
```
This should be inside the if statement. We don't want another row if there's no discord id.

BenjaminSchaaf commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -114,7 +115,7 @@
       expect(user.email).to eq('foo@bar.com')
       expect(user).to_not be_confirmed
-      expect(response).to redirect_to(user_path(user))
+      expect(response).to redirect_to(user_path(user)), type: :request
```
I don't think `type: :request` does anything here. That's something you'd normally see passed to `describe` so I'm not sure what it's even doing here?

AuroraEchoes replied:
I was messing around with request/routing specs for `be_routable`. So, best guess, I deleted the `type: request` from a describe block (which then went to my clipboard, because, `nvim`) and then carelessly middle clicked at some point when scrolling, which pasted that and I didn’t notice. (Very silly mistake to make and not notice, I apologise.)

BenjaminSchaaf commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -360,4 +361,50 @@
+        expect(user.discord_id).to_not be_nil
+        patch :unlink_discord, params: { id: user.id }
+        other.reload
+        expect(user.discord_id).to_not be_nil
```
If we're testing `user` here, then `user` needs to be reloaded. We should probably test both.

AuroraEchoes replied:
In hindsight, this is a weird way to test this. Updating to simply sign out the user, `patch :unlink_discord`, and confirm it doesn’t work.

BenjaminSchaaf commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -360,4 +361,50 @@
+
+    context 'with Discord integration disabled' do
+      before do
+        allow(Features).to receive(:discord_integration_enabled?).and_return(true)
```
The context is that the integration is disabled, so we should be disabling it.

BenjaminSchaaf reviewed (commented):
Some minor things left. I'd also appreciate if this was rebased onto master before I merge it.

AuroraEchoes committed:
Discord Integration unit test fixes

AuroraEchoes commented:
Done! Beware — all builds have failed since you added a workflow run to `pull_requests` — apparently it’s “not a valid event name” (according to the [docs](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request), it shouldn’t be pluralised).

AuroraEchoes committed:
Fix Discord link not being on it’s own line

AuroraEchoes committed:
Fix dependencies

AuroraEchoes committed:
Rubocop

AuroraEchoes committed:
Try to fix RuboCop

Shigbeard commented:
Hi @BenjaminSchaaf 

Can we please work to resolve this pull request quickly? We have a number of outstanding items in other projects that are dependent on these features making it to production.

Thank you.

BenjaminSchaaf committed:
Remove features.rb file; just access Rails.configuration directly

BenjaminSchaaf committed:
Add authentication check to discord logic

This avoids errors when a discord login is initiated when the user is
not logged in.
